numerics/polynomial_evaluators_body.hpp

```diff
@@ -118,104 +120,150 @@ struct InternalEstrinEvaluator<Value, Argument, degree, low, 0> {
       ArgumentSquares const& argument_squares);
 };
 
-template<typename Value, typename Argument, int degree, int low, int subdegree>
-Derivative<Value, Argument, low>
-InternalEstrinEvaluator<Value, Argument, degree, low, subdegree>::Evaluate(
-    Coefficients const& coefficients,
-    Argument const& argument,
-    ArgumentSquares const& argument_squares) {
+template<typename Value,
+         typename Argument,
+         int degree,
+         int low,
+         int subdegree,
+         bool fma>
+FORCE_INLINE(inline) Derivative<Value, Argument, low>
+InternalEstrinEvaluator<Value, Argument, degree, low, subdegree, fma>::
+    Evaluate(Coefficients const& coefficients,
+             Argument const& argument,
+             ArgumentSquares const& argument_squares) {
   static_assert(subdegree >= 2,
                 "Unexpected subdegree in InternalEstrinEvaluator::Evaluate");
   // |n| is used to select |argument^(2^(n + 1))| = |argument^m|.
   constexpr int n = FloorLog2(subdegree) - 1;
   // |m| is |2^(n + 1)|.
   constexpr int m = PowerOf2Le(subdegree);
-  return InternalEstrinEvaluator<Value, Argument, degree,
-                                 low, m - 1>::
-             Evaluate(coefficients, argument, argument_squares) +
-         std::get<n>(argument_squares) *
-             InternalEstrinEvaluator<Value, Argument, degree,
-                                     low + m, subdegree - m>::
-                 Evaluate(coefficients, argument, argument_squares);
+  auto const& xᵐ = std::get<n>(argument_squares);
+  auto const a = InternalEstrinEvaluator<Value,
+                                         Argument,
+                                         degree,
+                                         low + m,
+                                         subdegree - m,
+                                         fma>::Evaluate(coefficients,
+                                                        argument,
+                                                        argument_squares);
+  auto const b =
+      InternalEstrinEvaluator<Value, Argument, degree, low, m - 1, fma>::
+          Evaluate(coefficients, argument, argument_squares);
+  if constexpr (fma) {
```

Comment:
It would be tempting to pass a pointer-to-function instead of the disgraceful `fma` bit, and have a helper function to compute `a * b + c` in the non-FMA case.  Unfortunately, that's probably hard because writing the parameter types for that function would be hard.

Reply:
Yes, and naming the « multiply-and-add, maybe fused but maybe not » operation seems weird—the FMA calls are already ugly but at least here they parallel a nice a * b + c.